<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="cleaning_bot">
    <!-- Xacro property for color brightness and alpha. -->
    <xacro:property name="bright" value=".75" />
    <xacro:property name="alpha" value="1" />

    <!-- Xacro property for scale. -->
    <xacro:property name="scale" value="1" />

    <!-- Scale derived from 'scale' property. -->
    <xacro:property name="chassis_radius" value="${scale*2}" />
    <xacro:property name="chassis_length" value="${scale*1}" />
    <xacro:property name="wheel_radius" value="${chassis_radius/2}" />
    <xacro:property name="wheel_length" value="${chassis_length/2}" />
    <xacro:property name="caster_radius" value="${chassis_radius/4}" />
    <xacro:property name="lidar_length" value="${chassis_radius/4}" />

    <!-- List of materials for easier viewing. -->
    <material name="red">
        <color rgba="${bright} 0.0 0.0 ${alpha}"/>
    </material>

    <material name="blue">
        <color rgba="0 0 ${bright} ${alpha}" />
    </material>

    <material name="green">
        <color rgba="0 ${bright} 0 ${alpha}" />
    </material>

    <material name="white">
        <color rgba="${bright} ${bright} ${bright} ${alpha}"/>    
    </material>

    <material name="purple">
        <color rgba="${bright} 0 ${bright*0.75} ${alpha}" />
    </material>

    <!-- World link as the root. -->
    <link name="world"/>

    <!-- This is the base link of the robot. A simple cylinderical chassis. -->
    <link name="base_link">
        <visual>
           <geometry>
                <cylinder radius="${chassis_radius}" length="${chassis_length}"/> 
           </geometry>
           <material name="white"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${chassis_radius}" length="${chassis_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="10.0"/>
            <inertia ixx="10.8333" ixy="0.0" ixz="0.0" iyy="10.8333" iyz="0.0" izz="20.0"/>
        </inertial>
    </link>

    <!-- This is the left wheel link. -->
    <link name="left_wheel_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.2708" ixy="0.0" ixz="0.0" iyy="0.2708" iyz="0.0" izz="0.5"/>
        </inertial>
    </link>

    <!-- This is the right wheel link. -->
    <link name="right_wheel_link">
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <material name="red"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="1.0"/>
            <inertia ixx="0.2708" ixy="0.0" ixz="0.0" iyy="0.2708" iyz="0.0" izz="0.5"/>
        </inertial>
    </link>

    <!-- The caster wheel. -->
    <link name="caster_wheel_link">
        <visual>
            <geometry>
                <sphere radius="${caster_radius}"/>
            </geometry>
            <material name="green"/>
        </visual>
        <collision>
            <geometry>
                <sphere radius="${caster_radius}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.5"/>
            <inertia ixx="0.05" ixy="0.0" ixz="0.0" iyy="0.05" iyz="0.0" izz="0.05"/>
        </inertial>
    </link>

    <!-- This is the lidar camera link mounted on top of the chassis. -->
    <link name="lidar_link">
        <visual>
            <geometry>
                <box size="${lidar_length} ${lidar_length} ${lidar_length}"/>
            </geometry>
            <material name="purple"/>
        </visual>
        <collision>
            <geometry>
                <box size="${lidar_length} ${lidar_length} ${lidar_length}"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="0.2"/>
            <inertia ixx="0.004167" ixy="0.0" ixz="0.0" iyy="0.004167" iyz="0.0" izz="0.004167"/>
        </inertial>
    </link>

    <!-- This joint defines offset of base joint from world root. -->
    <joint name="world_base_joint" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
        <origin xyz="0.0 0.0 ${chassis_radius+caster_radius}" rpy="0.0 0.0 1.57" />
    </joint>

    <!-- This joint defines base/left-wheel joint. -->
    <joint name="base_left_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="left_wheel_link"/>
        <origin xyz="0 ${chassis_radius+wheel_length/2} -${chassis_length/2}" rpy="1.57 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <!-- This joint defines base/right-wheel joint. -->
    <joint name="base_right_wheel_joint" type="continuous">
        <parent link="base_link"/>
        <child link="right_wheel_link"/>
        <origin xyz="0 -${chassis_radius+wheel_length/2} -${chassis_length/2}" rpy="1.57 0 0"/>
        <axis xyz="0 0 1"/>
    </joint>

    <!-- This joint defines base/caster-wheel joint. -->
    <joint name="base_caster_wheel_joint" type="fixed">
        <parent link="base_link"/>
        <child link="caster_wheel_link"/>
        <origin xyz="0 0 -${chassis_length}" rpy="0.0 0.0 0.0"/>
    </joint>

    <!-- This joint defines base/lidar joint. -->
    <joint name="base_lidar_joint" type="fixed">
        <parent link="base_link"/>
        <child link="lidar_link"/>
        <origin xyz="${chassis_radius-lidar_length/2} 0 ${chassis_length-lidar_length/2}" rpy="0.0 0.0 0.0"/>
    </joint>

</robot>